sdk/include/opentelemetry/sdk/trace/simple_processor.h

```diff
@@ -22,24 +26,29 @@ class SimpleSpanProcessor : public SpanProcessor
    * @param exporter the exporter used by the span processor
    */
   explicit SimpleSpanProcessor(std::unique_ptr<SpanExporter> &&exporter) noexcept
-      : exporter_(std::move(exporter))
+      : exporter_(std::move(exporter)), lock_(ATOMIC_FLAG_INIT)
   {}
 
   std::unique_ptr<Recordable> MakeRecordable() noexcept override
   {
-    return exporter_->MakeRecordable();
+    WaitLock();
+    auto result = exporter_->MakeRecordable();
```

Comment:
We don't need to lock around `MakeRecodable`.

Reply:
Is that an implementation detail of the exporters?

I.e. does the exporter require `MakeRecordable` to be threadsafe, while `Export` + `Shutdown` are not?

For reference only `Export` is called out as requiring one-thread-access in the [code comments](https://github.com/open-telemetry/opentelemetry-cpp/blob/master/sdk/include/opentelemetry/sdk/trace/exporter.h).

Should I update the docs on Exporter to denote (a) `MakeRecordable` needs to be thread-safe and (b) `Shutdown` will only be called on on thread?

Reply:
1. According to [the spec](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/sdk.md#exportbatch), "Export() will never be called concurrently for the same exporter instance". That's why we need the lock around the call to `Export`.
2. The spec doesn't specify any concurrency requirements around `Shutdown`. So we could decide (and document) whether the `Shutdown` functions on the exporter should be re-entrant or not.
3. `MakeRecordable` isn't required by the spec, but is provided by the C++ SDK due to an architecture which focuses on performance. Every exporter has to implement `MakeRecordable` in a way so that it can be called concurrently from different threads (performance, again). It would be great if you could add that requirement to the documentation of `MakeRecordable`.

So: we must assume that `Export` is not thread-safe, because of the spec. We assume that `MakeRecordable` is thread-safe, for performance reasons. For `Shutdown` it's up for us to decide. I have no strong preference there.

Reply:
As @pyohannes mentioned - if there is more than one call to the `Shutdown` function, the behavior is undefined by the spec.

Regarding our choice for OpenTelemetry C++, this is something we can borrow https://github.com/open-telemetry/opentelemetry-dotnet/blob/8ccabb324c8f51609c78df6651c165a7902cf009/src/OpenTelemetry/BaseProcessor.cs#L123.

Reply:
I didn't update anything around shutdown yet, but I'm going to add some at-least-once code.  Is there an open bug about that or should I just do it in this PR?

Reply:
Added it in this PR.